sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java

```diff
@@ -2830,12 +2832,21 @@ private List<Document> flushBatch() throws IOException, InterruptedException {
 
         HttpEntity requestBody =
             new NStringEntity(bulkRequest.toString(), ContentType.APPLICATION_JSON);
+
+        String elasticResponseExceptionMessage = null;
+
         try {
           Request request = new Request("POST", endPoint);
           request.addParameters(Collections.emptyMap());
           request.setEntity(requestBody);
           response = restClient.performRequest(request);
           responseEntity = new BufferedHttpEntity(response.getEntity());
+        } catch (ResponseException ex) {
+          if (spec.getThrowWriteErrors()) {
+            throw ex;
+          } else {
+            elasticResponseExceptionMessage = ex.getMessage();
+          }
```

Comment:
I think this section should be integrated below within the IOException handler. Because `ResponseException` inherits from IOException[1], many exceptions that are currently going through the retry code path might instead go through this new path.  Moving this path under the IOException handler will provide broader coverage of failure modes where throwing the exception can be avoided.  

Within the block after the predicate `if (spec.getRetryConfiguration() == null || !isRetryableClientException(ex))`, this section could replace the existing line `throw ex;` (currently line 2852)

[1] https://artifacts.elastic.co/javadoc/org/elasticsearch/client/elasticsearch-rest-client/7.17.0/org/elasticsearch/client/ResponseException.html